### DISCUSSION TITLE
Route definition in bundles

### DISCUSSION THREAD

**Discussion by ofrzeta** [symfony/symfony#49168] [Q&A] [answered]
Hi, I want to start using bundles in Symfony 6 to enable a development workflow were we create reusable modules that can be shared between customer sites. Unfortunately I can hardly find any info about Symfony 6 bundles besides the official docs. Other resources on the Internet seem out of date as the bundle structure has changed if I understand correctly.

Now one use case is to "override" some routes that are defined in the base app with corresponding routes in a bundle. Let's say we have a src/Controller/Admin/DashboardController.php that defines a route in the attribute "#[Route('/admin', name: 'admin')]". Now I have a corresponding controller in our own package (lib/CompanyAdminBundle/src/Controller/Admin/DashboardController.php) with an annotation #[Route('/admin', name: 'admin')]". That route doesn't seem to be picked up by the system, even if I provide a "priority: 100". 

What am I doing wrong and what's the right way to make this work?

Thanks
Oliver

**Comment by ahundiak** [comment, marked as answer]
Couple of things to understand about bundles and routing.

The first is that bundle routes are never loaded automatically.  You always need a bit of configuration at the application level.  The docs can be a bit confusing on this point but typically you would place a file under config/routes:
```
# project/config/routes/cerad_my.yaml
cerad_my_controllers:
    resource:
        path: '@CeradMyBundle/src/Controller/'
        namespace: Cerad\MyBundle\Controller
    type: attribute
    prefix: cerad_my
```
It is a bit unusual to use attributes/annotations in bundles.  Defining your routes in one file is recommended but it works.  I'd suggest making sure you can define one bundle route before you worry about overriding stuff. `bin/console debug:router` is your friend here.

By the way, route priority controls the order in which routes are matched.  Nothing to do with overriding an actual route.

The next thing is that the way to control overriding an existing route is to control the order in which routes are loaded.  Last route of a given name will win.  Normally of course you want the application to be able to override bundle routes.  It's considered a bit rude to override application routes.  However since config/routes.yaml is by default loaded last (see App\Kernel::configureRoutes()), you can load your bundle routes at the end of routes.yaml and they will override anything else.
